kafka/src/main/java/com/metamx/tranquility/kafka/KafkaConsumer.java

```diff
@@ -0,0 +1,273 @@
+          countsSinceLastCommit.isEmpty() ? "0" : countsSinceLastCommit,
+          commitStartTime - flushStartTime,
+          finishedTime - commitStartTime
+      );
+    }
+    finally {
+      commitLock.writeLock().unlock();
+    }
+  }
+
+  private Runnable createCommitRunnable()
+  {
+    return new Runnable()
+    {
+      @Override
+      public void run()
+      {
+        long lastFlushTime = System.currentTimeMillis();
+        try {
+          while (!Thread.currentThread().isInterrupted()) {
+            Thread.sleep(Math.max(commitMillis - (System.currentTimeMillis() - lastFlushTime), 0));
+            commit();
+            lastFlushTime = System.currentTimeMillis();
+          }
+        }
+        catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          log.info(e, "Commit thread interrupted.");
+        }
+        catch (Throwable e) {
+          log.error(e, "Commit thread failed!");
+          throw Throwables.propagate(e);
+        }
+        finally {
+          stop();
+        }
+      }
+    };
+  }
+
+  private void startConsumers()
+  {
+    final List<KafkaStream<byte[], byte[]>> kafkaStreams = consumerConnector.createMessageStreamsByFilter(
+        topicFilter,
+        numThreads
+    );
+
+    for (final KafkaStream<byte[], byte[]> kafkaStream : kafkaStreams) {
+      consumerExec.submit(
+          new Runnable()
+          {
+            @Override
+            public void run()
+            {
+              try {
+                final Iterator<MessageAndMetadata<byte[], byte[]>> kafkaIterator = kafkaStream.iterator();
+
+                while (kafkaIterator.hasNext()) {
+                  if (Thread.currentThread().isInterrupted()) {
+                    throw new InterruptedException();
+                  }
+
+                  // Kafka consumer treats messages as consumed and updates in-memory last offset when the message
+                  // is returned by next(). In order to guarantee at-least-once message delivery, we need to a) set
+                  // autocommit enable to false so the consumer will not automatically commit offsets to Zookeeper, and
+                  // b) synchronize calls of kafkaIterator.next() with the commit thread so that we don't read messages
+                  // and then call consumerConnector.commitOffsets() before those messages have been flushed through
+                  // Tranquility into the indexing service.
+                  commitLock.readLock().lockInterruptibly();
+
+                  try {
+                    MessageAndMetadata<byte[], byte[]> data = kafkaIterator.next();
+                    writerController.getWriter(data.topic()).send(data.message());
+                  }
+                  finally {
+                    commitLock.readLock().unlock();
+                  }
+                }
+              }
+              catch (InterruptedException e) {
+                log.info(e, "Consumer thread interrupted.");
```

Comment:
Similar comment about the `e, ` (interrupting is going to happen on normal shutdowns)